src/audio/mixin_mixout.c

```diff
@@ -879,39 +855,43 @@ static int mixout_copy(struct comp_dev *dev)
 	list_for_item(blist, &dev->bsource_list) {
 		struct comp_buffer *unused_in_between_buf;
 		struct comp_dev *mixin;
-		struct mixout_source_info *src_info;
+		struct mixin_data *mixin_data;
 
 		unused_in_between_buf = buffer_from_list(blist, struct comp_buffer,
 							 PPL_DIR_UPSTREAM);
 		mixin = buffer_get_comp(unused_in_between_buf, PPL_DIR_UPSTREAM);
+		mixin_data = comp_get_drvdata(mixin);
 
-		src_info = find_mixout_source_info(mixed_data_info, mixin);
-		/* this shouldn't happen but skip even if it does and move to the next source */
-		if (!src_info)
+		if (mixin->state != COMP_STATE_ACTIVE)
 			continue;
 
 		/* Inactive sources should not block other active sources */
 		if (comp_get_state(dev, mixin) == COMP_STATE_ACTIVE)
 			frames_to_produce = MIN(frames_to_produce,
-						src_info->consumed_yet_not_produced_frames);
+						mixin_data->consumed_yet_not_produced_frames);
 	}
 
 	if (frames_to_produce > 0 && frames_to_produce < INT32_MAX) {
-		int i;
 		struct comp_buffer *sink;
 		struct comp_buffer __sparse_cache *sink_c;
 
-		for (i = 0; i < MIXOUT_MAX_SOURCES; i++) {
-			if (!mixed_data_info->source_info[i].mixin)
+		list_for_item(blist, &dev->bsource_list) {
+			struct comp_buffer *unused_in_between_buf;
+			struct mixin_data *mixin_data;
+			struct comp_dev *mixin;
+
+			unused_in_between_buf = buffer_from_list(blist, struct comp_buffer,
+								 PPL_DIR_UPSTREAM);
+			mixin = buffer_get_comp(unused_in_between_buf, PPL_DIR_UPSTREAM);
+			mixin_data = comp_get_drvdata(mixin);
+
+			if (mixin->state != COMP_STATE_ACTIVE)
```

Comment:
Here indeed I think problems are possible, since this is a `.copy()` method so IPCs can run in parallel, so mixins can get added or removed while this code executes